helix-term/src/ui/explore.rs

```diff
@@ -0,0 +1,959 @@
+    Filter,
+}
+
+#[derive(Clone, Debug)]
+struct State {
+    focus: bool,
+    open: bool,
+    current_root: PathBuf,
+    area_width: u16,
+    filter: String,
+}
+
+impl State {
+    fn new(focus: bool, current_root: PathBuf) -> Self {
+        Self {
+            focus,
+            current_root,
+            open: true,
+            area_width: 0,
+            filter: "".to_string(),
+        }
+    }
+}
+
+pub struct Explorer {
+    tree: TreeView<FileInfo>,
+    history: Vec<TreeView<FileInfo>>,
+    show_help: bool,
+    state: State,
+    prompt: Option<(PromptAction, Prompt)>,
+    #[allow(clippy::type_complexity)]
+    on_next_key: Option<Box<dyn FnMut(&mut Context, &mut Self, &KeyEvent) -> EventResult>>,
+    #[allow(clippy::type_complexity)]
+    repeat_motion: Option<Box<dyn FnMut(&mut Self, PromptAction, &mut Context) + 'static>>,
+    column_width: u16,
+}
+
+impl Explorer {
+    pub fn new(cx: &mut Context) -> Result<Self> {
+        let current_root = std::env::current_dir().unwrap_or_else(|_| "./".into());
+        Ok(Self {
+            tree: Self::new_tree_view(current_root.clone())?,
+            history: vec![],
+            show_help: false,
+            state: State::new(true, current_root),
+            repeat_motion: None,
+            prompt: None,
+            on_next_key: None,
+            column_width: cx.editor.config().explorer.column_width as u16,
+        })
+    }
+
+    fn new_tree_view(root: PathBuf) -> Result<TreeView<FileInfo>> {
+        let root = FileInfo::root(root.clone());
+        let children = root.get_children()?;
+        Ok(TreeView::build_tree(root, children).with_enter_fn(Self::toggle_current))
+    }
+
+    fn push_history(&mut self, tree_view: TreeView<FileInfo>) {
+        self.history.push(tree_view);
+        const MAX_HISTORY_SIZE: usize = 20;
+        Vec::truncate(&mut self.history, MAX_HISTORY_SIZE)
+    }
+
+    fn change_root(&mut self, cx: &mut Context, root: PathBuf) {
+        if self.state.current_root.eq(&root) {
+            return;
+        }
+        match Self::new_tree_view(root.clone()) {
+            Ok(tree) => {
+                let old_tree = std::mem::replace(&mut self.tree, tree);
+                self.push_history(old_tree);
+                self.state.current_root = root;
+            }
+            Err(e) => cx.editor.set_error(format!("{e}")),
+        }
+    }
+
+    fn reveal_file(&mut self, path: PathBuf) -> Result<()> {
+        let current_root = &self.state.current_root;
+        let current_path = path.as_path().to_string_lossy().to_string();
+        let current_root = current_root.as_path().to_string_lossy().to_string() + "/";
+        let segments = current_path
+            .strip_prefix(current_root.as_str())
+            .expect(
+                format!(
+                    "Failed to strip prefix '{}' from '{}'",
+                    current_root, current_path
+                )
+                .as_str(),
+            )
```

Comment:
Just a heads, this panics on opening the explorer if you open any file from outside the current working directory, such as doing `hx ../foo` or `hx /path/to/foo`.

Not sure what the best workaround for this is, things like the file picker still root from the current working directory, so maybe this function should skip on files not located in the current root?

Great work so far, looking forward to this!